### DISCUSSION TITLE
SCV with partner telephony, Users created with alias@userid@orgid issues with Extension and vmx2

### DISCUSSION THREAD

**Discussion by SyndrixLane** [amazon-connect/amazon-connect-salesforce-scv#122] [Q&A] [answered]
Hello,

I am setting up SCV with partner telephony with amazon connect. This is all new to me.

When I add a user to my contact center in salesforce, it creates the user in the AWS connect instance with alias@userId@orgId

This caused me issues with the extension router and the VMX2 solution.

VMX2 issue only with agent, queue delivery worked.

I made a temporary fix in the agent voicemail delivery but I would appreciate if someone can show me how I can clean this code up.  The original code was passing the agent with alias@userid@orgid so it was failing

This might not be the right way to fix this, but I found that in the contact attributes the extension they dialed was still there, since the only way to get direct to agent voicemail is using extension in my deployment I edited the code to use the extension to find the user ID
 
The small edits I made to the code that is there to make it work, this plugs the extension into the sf_agent_id value which then fails the "005" check which then allows me to query using the extension to pull the right id

         # Validate the agent's Salesforce ID, and correct, if necessary
            sf_agent_id = writer_payload['json_attributes']['extension']

            if not sf_agent_id.startswith('005'):
                try:
                    get_agent_id = sf.query("SELECT Id FROM User WHERE Extension = '" + sf_agent_id + "' LIMIT 1")
                    sf_agent_id = get_agent_id[0]['Id']
                except Exception as e:
                    logger.error(e)
                    logger.error('Record {0} Result: Failed to validate Salesforce User ID'.format(writer_payload['loop_counter']))
                    return 'fail'
            if '@' in sf_agent_id:
                sf_agent_id = sf_agent_id.split('@')[0]

I would like to improve/shorten this code, but I do not know how to code so when I try it fails... This is what I tried.
# Validate the agent's Salesforce ID, and correct, if necessary
  sf_agent_ext = writer_payload['json_attributes']['extension']
get_agent_id = sf.query("SELECT Id FROM User WHERE Extension = '" sf_agent_ext "' ")
sf_agent_id = get_agent_id[0]['Id']
 except Exception as e:
                    logger.error(e)
                    logger.error('Record {0} Result: Failed to validate Salesforce User ID'.format(writer_payload['loop_counter']))
                    return 'fail'

Extension Router issue and resolution I made

 
The first lambda  block in the extension router flow I expanded the query
SELECT User.Id, User.Alias FROM User WHERE Extension = '$.Attributes.extension'

The second lambda block I saved the additional alias 
Use attribute
Destination key
Id
Type
External
Attribute
Id
Use text

Use attribute
Destination key
Alias
Type
External
Attribute
Alias

In the extension routing lambda I updated the code so it would compare the full alias@userid@orgid
   # Extract the passed Id attribute, if present
    try:
        Id = event['Details']['Parameters']['Id']
        Alias = event['Details']['Parameters']['Alias']

        # If the key exists, extract the value and build the agent id
        if Id != '':
            clean_id = Id[:- 3]
            agent_id = Alias + '@' + clean_id + '@' + org_id
            response.update({'agent_id':agent_id})

This seems to work fine with these changes.

Thank you

**Comment by dougjaso** [comment, marked as answer]
Pushed a fix for this for VMX2 about 20 minutes ago. Problem is that there are 2 versions of the agent ID for SCV depending on when you deployed, and that the same code needs to work for non-scv deployments as well. I'll also implement something similar for Extension routing.

If you are using cases, you can see the new code here: https://github.com/amazon-connect/amazon-connect-salesforce-scv/blob/master/Solutions/VMX2-VoicemailExpress/Code/sub_salesforce_case.py

if you are using custom objects, you can see the new code here: https://github.com/amazon-connect/amazon-connect-salesforce-scv/blob/master/Solutions/VMX2-VoicemailExpress/Code/sub_salesforce_other.py

**Reply by SyndrixLane** [reply]
Thank you!
